junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/MethodBasedCondition.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.api.condition;
+
+import static java.lang.String.format;
+
+import java.lang.reflect.Method;
+import java.util.Optional;
+
+import org.junit.jupiter.api.extension.ConditionEvaluationResult;
+import org.junit.jupiter.api.extension.ExecutionCondition;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.platform.commons.JUnitException;
+import org.junit.platform.commons.util.Preconditions;
+import org.junit.platform.commons.util.ReflectionUtils;
+
+public abstract class MethodBasedCondition implements ExecutionCondition {
+
+	abstract Optional<String> getMethodName(ExtensionContext context);
+
+	abstract ConditionEvaluationResult getDefaultResult();
+
+	abstract ConditionEvaluationResult getResultBasedOnBoolean(boolean result);
+
+	@Override
+	public ConditionEvaluationResult evaluateExecutionCondition(ExtensionContext context) {
+		return getMethodName(context) //
+				.map(methodName -> getConditionMethod(methodName, context)) //
+				.map(method -> (boolean) evaluateCondition(method, context)) //
+				.map(this::getResultBasedOnBoolean).orElse(getDefaultResult());
+	}
+
+	private Method getConditionMethod(String methodName, ExtensionContext context) {
+		if (methodName.contains("#")) {
+			return findMethodByFullyQualifiedName(methodName);
+		}
+		return findMethod(context.getRequiredTestClass(), methodName);
+	}
+
+	private Method findMethodByFullyQualifiedName(String fullyQualifiedMethodName) {
+		String[] methodParts = ReflectionUtils.parseFullyQualifiedMethodName(fullyQualifiedMethodName);
+		String className = methodParts[0];
+		String methodName = methodParts[1];
+		Class<?> clazz = ReflectionUtils.tryToLoadClass(className).getOrThrow(
+			cause -> new JUnitException(format("Could not load class [%s]", className), cause));
+		return findMethod(clazz, methodName);
+	}
+
+	private Method findMethod(Class<?> clazz, String methodName) {
+		return ReflectionUtils.findMethod(clazz, methodName) //
+				.orElseGet(() -> ReflectionUtils.findMethod(clazz, methodName, ExtensionContext.class) //
+						.orElseThrow(() -> new JUnitException(
+							format("Could not find method [%s] in class [%s]", methodName, clazz.getName()))));
```

Comment:
Can we reuse this logic from `MethodArgumentsProvider` somehow?

Reply:
Hmm I thought about that at first, but since the classes were two different modules I put that aside for now. But I agree.
Where would you put the logic in common? In `ReflectionUtils`?

Reply:
> Where would you put the logic in common? In `ReflectionUtils`?

Sounds like a good place. We may export that logic to external users later, if requested.